docs: add tokenfactory tutorial (#3730)
ignite/cli: docs/docs/02-guide/06-tokenfactory/01-tokenfactory.md

# Token Factory

## Introduction to Building a Token Factory Module with Ignite CLI

In this tutorial, we will guide you through the process of building a token factory module using the Ignite CLI. This module is a powerful tool for creating native denominations (denoms) on your blockchain, providing you with the capability to issue and manage digital assets natively within your network.

> Digital assets, characterized by their uniqueness and scarcity, are fundamental to the value proposition of blockchain technology. A well-known example is the ERC20 standard on Ethereum, which has gained widespread popularity. By learning to create and manage native denoms on your blockchain, you will gain hands-on experience with one of blockchain's key functionalities.

Reviewer: ```suggestion
Digital assets, characterized by their uniqueness and scarcity, are fundamental to the value proposition of blockchain technology. A well-known example is the ERC20 standard on Ethereum, which has gained widespread popularity. By learning to create and manage native denoms on your blockchain, you will gain hands-on experience with one of the blockchain's key functionalities.
```

**You will learn how to:**

* Develop a module from scratch.
* Implement a CRUD (Create, Read, Update, Delete) operation while specifically removing the delete functionality to safeguard the integrity of initialized denoms.
* Integrate logic for creating new denoms.

> * Engage with various components such as the client, types, keeper, expected keeper, and handlers to effectively implement the Token Factory module.

Reviewer: ```suggestion
* Engage with various components such as the client, types, keeper, expected keeper, and handlers to implement the Token Factory module effectively.
```

```bash
ignite scaffold chain tokenfactory --no-module
```

This command establishes a new Cosmos SDK blockchain named `tokenfactory` and places it in a directory of the same name. Inside this directory, you'll find a fully functional blockchain ready for further customization.

Now, navigate into your newly created blockchain directory:

```bash
cd tokenfactory
```

### Scaffold Your Token Factory Module

Next, we'll scaffold a new module for your token factory. This module will depend on the Cosmos SDK's [bank](https://docs.cosmos.network/main/build/modules/bank#abstract) and [auth](https://docs.cosmos.network/main/build/modules/auth#abstract) modules, which provide essential functionalities like account access and token management. Use the following command:

```bash
ignite scaffold module tokenfactory --dep account,bank
```

The successful execution of this command will be confirmed with a message indicating that the `tokenfactory` module has been created.

### Defining Denom Data Structure

> To manage denoms within your token factory, define their structure using a Ignite map. This will store the data as key-value pairs. Run this command:

Reviewer: ```suggestion
To manage denoms within your token factory, define their structure using an Ignite map. This will store the data as key-value pairs. Run this command:
```

```bash
ignite scaffold map Denom description:string ticker:string precision:int url:string maxSupply:int supply:int canChangeMaxSupply:bool --signer owner --index denom --module tokenfactory
```

> Review the `proto/tokenfactory/tokenfactory/denom.proto` file to see the scaffolding results, which include modifications to various files indicating successful creation of the denom structure.

Reviewer: ```suggestion
Review the `proto/tokenfactory/tokenfactory/denom.proto` file to see the scaffolding results, which include modifications to various files indicating the successful creation of the denom structure.
```

### Git Commit

After scaffolding your denom map, it's a good practice to save your progress. Use the following commands to make your first Git commit:

```bash
git add .
git commit -m "Add tokenfactory module and denom map"
```

This saves a snapshot of your project, allowing you to revert back if needed.

## Removing Delete Functionality

In a blockchain context, once a denom is created, it's crucial to ensure it remains immutable and cannot be deleted. This immutability is key to maintaining the integrity and trust in the blockchain. Therefore, we'll remove the delete functionality from the scaffolded CRUD operations. Follow these steps:

**Proto Adjustments**

In `proto/tokenfactory/tokenfactory/tx.proto`, remove the `DeleteDenom` RPC method and the associated message types.

**Client Updates**

> Navigate to the client in `x/tokenfactory/client` and make these changes:

Reviewer: The client package will change in the new Ignite version with the cosmos-sdk v0.50. We will use only the autocli

Author: Ok, let's update when the new version arrives.

- Remove `TestDeleteDenom()` from `tx_denom_test.go`.
- Eliminate `CmdDeleteDenom()` from `tx_denom.go`.
- In `tx.go`, delete the line referencing the delete command.

**Keeper Modifications**

In `denom_test.go`, remove `TestDenomRemove()`.
Delete `RemoveDenom()` from `denom.go`.
Exclude `TestDenomMsgServerDelete()` and `DeleteDenom()` functions from `msg_server_denom_test.go` and `msg_server_denom.go`, respectively.

**Types Directory Changes**

- Update `codec.go` to remove references to `MsgDeleteDenom`.
- Remove `TestMsgDeleteDenom_ValidateBasic()` from `messages_denom_test.go`.
- Eliminate all references to `MsgDeleteDenom()` in `messages_denom.go`.

After making these changes, commit your updates:

```bash
git add .
git commit -m "Remove the delete denom functionality"
```

> This concludes the second chapter, setting a solid foundation for your token factory module. In the next chapter, we'll delve into implementing the application logic that will bring your token factory to life.

Reviewer: ```suggestion
This concludes the second chapter, setting a solid foundation for your token factory module. In the next chapter, we'll delve into implementing the application logic to bring your token factory to life.
```

**Proto Definition Updates**

Start by defining the structure of a new token denom in `proto/tokenfactory/tokenfactory/tx.proto`.

For `MsgCreateDenom`:

- Remove `int32 supply = 8;` and adjust the field order so `canChangeMaxSupply` becomes the 8th field.

Resulting `MsgCreateDenom` message:

```proto
message MsgCreateDenom {
  string owner = 1;
  string denom = 2;
  string description = 3;
  string ticker = 4;
  int32 precision = 5;
  string url = 6;
  int32 maxSupply = 7;
  bool canChangeMaxSupply = 8;
}
```

For `MsgUpdateDenom`:

- Omit `string ticker = 4;`, `int32 precision = 5;`, and `int32 supply = 8;`, and reorder the remaining fields.

Resulting `MsgUpdateDenom` message:

```proto
message MsgUpdateDenom {
  string owner = 1;
  string denom = 2;
  string description = 3;
  string url = 4;
  int32 maxSupply = 5;
  bool canChangeMaxSupply = 6;
}
```

### Client Logic

In the `x/tokenfactory/client/cli/tx_denom.go` file, update the client application logic.

**For `CmdCreateDenom`:**

> - Adjust the number of arguments from 8 to 7, removing references to the supply argument, and update the usage descriptions.

Reviewer: maybe it is better to add an example below the `Client Logic` explanation

Author: This is removing one line and editing another, the `supply` - does this really need the code?

**For `CmdUpdateDenom()`:**

- Reduce the number of arguments to 5, excluding `supply`, `precision`, and `ticker`, and modify the usage descriptions accordingly.

Also, update the tests in `x/tokenfactory/client/cli/tx_denom_test.go` to reflect these changes.

### Types Updates

When creating new denoms, they initially have no supply. The supply is determined only when tokens are minted.

In `x/tokenfactory/types/messages_denom.go`:

- Remove the `supply` parameter from `NewMsgCreateDenom`.
- Update `NewMsgUpdateDenom` to exclude unchangeable parameters like `ticker`, `precision`, and `supply`.

Implement basic input validation in `x/tokenfactory/types/messages_denom.go`:

- Ensure the ticker length is between 3 and 10 characters.
```go
func (msg *MsgCreateDenom) ValidateBasic() error {

>     _, err := sdk.AccAddressFromBech32(msg.Owner)
>     if err != nil {
>         return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid owner address (%s)", err)
>     }

Reviewer: `ErrInvalidAddress` will be deprecated

```suggestion
    if _, err := sdk.AccAddressFromBech32(m.Authority); err != nil {
        return sdkerrors.Wrap(err, "invalid authority address")
    }    
```

Author: Ok, let's update when the new version arrives.

    tickerLength := len(msg.Ticker)
    if tickerLength < 3 {
        return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Ticker length must be at least 3 chars long")
    }
    if tickerLength > 10 {

>         return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Ticker length must be 10 chars long maximum")

Reviewer: also, the other errors var were removed from the `cosmossdk.io/errors` pkg

Author: Is this already in place or only in the new version?

Reviewer: yes, some errors already show as deprecated

    }
    if msg.MaxSupply == 0 {
        return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Max Supply must be greater than 0")
    }
    
    return nil
}
```

- Set `maxSupply` to be greater than 0.

```go
func (msg *MsgUpdateDenom) ValidateBasic() error {
    _, err := sdk.AccAddressFromBech32(msg.Owner)
    if err != nil {
        return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid owner address (%s)", err)
    }
    if msg.MaxSupply == 0 {
        return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Max Supply must be greater than 0")
    }
    return nil
}
```

### Keeper Logic

The keeper is where you define the business logic for manipulating the database and writing to the key-value store.

**In `x/tokenfactory/keeper/msg_server_denom.go`:**

- Update `CreateDenom()` to include logic for creating unique denoms. Modify the error message to point to existing denoms. Set `Supply` to `0`.
- Modify `UpdateDenom()` to verify ownership and manage max supply changes.

```go
func (k msgServer) UpdateDenom(goCtx context.Context, msg *types.MsgUpdateDenom) (*types.MsgUpdateDenomResponse, error) {

>     ctx := sdk.UnwrapSDKContext(goCtx)

Reviewer: I think we don't need to wrap and unwrap context anymore in the new version of the sdk

### Scaffolding New Messages

**MintAndSendTokens:**

This message allows the creation (minting) of new tokens and their allocation to a specified recipient. The necessary inputs are the denom, the amount to mint, and the recipient's address.

Scaffold this message with:

```bash
ignite scaffold message MintAndSendTokens denom:string amount:int recipient:string --module tokenfactory --signer owner
```

**UpdateOwner:**

This message facilitates the transfer of ownership of a denom. It requires the denom name and the new owner's address.

Scaffold this message with:

```bash
ignite scaffold message UpdateOwner denom:string newOwner:string --module tokenfactory --signer owner
```

### Implementing Logic for New Messages

**In the `MintAndSendTokens` Functionality:**

Located in `x/tokenfactory/keeper/msg_server_mint_and_send_tokens.go`, this function encompasses the logic for minting new tokens. Key steps include:

- Verifying the existence and ownership of the denom.
- Ensuring minting does not exceed the maximum supply.

> - Minting the specified amount and sending it to the recipient.

Reviewer: ```suggestion
- Mint the specified amount and send it to the recipient.
```

```

- Check the recipient’s balance:
```bash
tokenfactoryd query bank balances cosmos16x46rxvtkmgph6jnkqs80tzlzk6wpy6ftrgh6t
```

- Verify the updated supply in denom list:
```bash
tokenfactoryd query tokenfactory list-denom
```

**5. Transferring Ownership:**

- Transfer the ownership of uignite:
```bash
tokenfactoryd tx tokenfactory update-owner uignite cosmos16x46rxvtkmgph6jnkqs80tzlzk6wpy6ftrgh6t --from alice
```

- Confirm the ownership change:
```bash
tokenfactoryd query tokenfactory list-denom
```

**6. Confirming Minting Restrictions:**

- Test minting with alice to ensure restrictions apply:

```bash
tokenfactoryd tx tokenfactory mint-and-send-tokens uignite 1200 cosmos16x46rxvtkmgph6jnkqs80tzlzk6wpy6ftrgh6t --from alice
```

## Congratulations!

You've successfully built and tested a token factory module. This advanced tutorial has equipped you with the skills to:

- Integrate other modules and utilize their functionalities.
- Customize CRUD operations to fit your blockchain's needs.
- Scaffold modules and messages effectively.

## Looking Ahead: IBC Functionality

> As you progress, the next learning adventure involves exploring IBC (Inter-Blockchain Communication). If you're up for a challenge, consider adding IBC functionality to your token factory module. This will not only enhance your module's capabilities but also deepen your understanding of the Cosmos ecosystem.

Reviewer: ```suggestion
As you progress, the following learning adventure involves exploring IBC (Inter-Blockchain Communication). If you're up for a challenge, consider adding IBC functionality to your token factory module. This will not only enhance your module's capabilities but also deepen your understanding of the Cosmos ecosystem.
```
